oshdb-api/src/main/java/org/heigit/bigspatialdata/oshdb/api/mapreducer/backend/Kernels.java

```diff
@@ -0,0 +1,266 @@
+import org.heigit.bigspatialdata.oshdb.api.object.OSMContribution;
+import org.heigit.bigspatialdata.oshdb.api.object.OSMEntitySnapshot;
+import org.heigit.bigspatialdata.oshdb.grid.GridOSHEntity;
+import org.heigit.bigspatialdata.oshdb.util.celliterator.CellIterator;
+
+class Kernels {
+  interface CellProcessor<S> extends SerializableBiFunction<GridOSHEntity, CellIterator, S> {}
+
+  interface CancelableProcessStatus {
+    boolean isActive();
+  }
+
+  private static class NonCancelableProcessStatus
+      implements CancelableProcessStatus, Serializable {
+    @Override
+    public boolean isActive() {
+      return true;
+    }
+  }
+
+  private static final CancelableProcessStatus NC = new NonCancelableProcessStatus();
+
+  // === map-reduce processors ===
+
+  @Nonnull
+  static <R, S> CellProcessor<S> getOSMContributionCellReducer(
+      SerializableFunction<OSMContribution, R> mapper,
+      SerializableSupplier<S> identitySupplier,
+      SerializableBiFunction<S, R, S> accumulator
+  ) {
+    return getOSMContributionCellReducer(mapper, identitySupplier, accumulator, NC);
+  }
+
+  @Nonnull
+  static <R, S> CellProcessor<S> getOSMContributionCellReducer(
+      SerializableFunction<OSMContribution, R> mapper,
+      SerializableSupplier<S> identitySupplier,
+      SerializableBiFunction<S, R, S> accumulator,
+      CancelableProcessStatus aborter
+  ) {
+    return (oshEntityCell, cellIterator) -> {
+      // iterate over the history of all OSM objects in the current cell
+      AtomicReference<S> accInternal = new AtomicReference<>(identitySupplier.get());
+      cellIterator.iterateByContribution(oshEntityCell)
+          .filter(ignored -> aborter.isActive())
+          .forEach(contribution -> {
+            OSMContribution osmContribution = new OSMContribution(contribution);
+            accInternal.set(accumulator.apply(accInternal.get(), mapper.apply(osmContribution)));
+          });
+      return accInternal.get();
+    };
+  }
+
+  @Nonnull
+  static <R, S> CellProcessor<S> getOSMContributionGroupingCellReducer(
+      SerializableFunction<List<OSMContribution>, Iterable<R>> mapper,
+      SerializableSupplier<S> identitySupplier,
+      SerializableBiFunction<S, R, S> accumulator
+  ) {
+    return getOSMContributionGroupingCellReducer(mapper, identitySupplier, accumulator, NC);
+  }
+
+  @Nonnull
+  static <R, S> CellProcessor<S> getOSMContributionGroupingCellReducer(
+      SerializableFunction<List<OSMContribution>, Iterable<R>> mapper,
+      SerializableSupplier<S> identitySupplier,
+      SerializableBiFunction<S, R, S> accumulator,
+      CancelableProcessStatus aborter
+  ) {
+    return (oshEntityCell, cellIterator) -> {
+      AtomicReference<S> accInternal = new AtomicReference<>(identitySupplier.get());
+      // iterate over the history of all OSM objects in the current cell
+      List<OSMContribution> contributions = new ArrayList<>();
```

Comment:
btw: while testing, I found that here (and in a few places below) re-using this array for multiple consecutive entities can cause a problem, when one just directly uses these arrays as output, e.g. the following would produce different results:

```java
OSMContributionView.….groupByEntity().collect() // <- wrong result
OSMContributionView.….groupByEntity().map(list -> new ArrayList(list)).collect()
```

This is kind of a corner case, since (at first glance) it only happens when one does not specify and map function and does not aggregate the results. But it could still be confusing and (I think) it doesn't matter too much from a performance point of view. I guess we should just not re-use the list here, or what do you think?